### ISSUE TITLE
enhancing @parallel and pmap

### ISSUE THREAD

**Issue by amitmurthy** [JuliaLang/julia#5500]
Starting a discussion around enhancing `@parallel` and `pmap` in the following ways.

1) `pmap` to take an additional keyword argument specifying the processes to be used. Defaults to all workers() as it currently is.

2)  `@parallel` has two variants currently 

        @parallel reducer loop
        @parallel loop

Suggesting additionally supporting the following

     @parallel procs loop
     @parallel reducer procs loop
     @parallel procs reducer  loop

where `procs` is a list of processes that `@parallel` should operate on. 

This will be quite useful to initialize SharedArrays in parallel (when it is present only on a subset of the workers).

For example

`@parallel procs(S) for i in 1:length(S); S[i] = rand(); end` where `S` is a `SharedArray`

I am not sure how we can differentiate between `procs` and `reducer` given that `procs` itself may be a result of an expression.

3) `@parallel` has an undocumented feature, `@parallel comprehension` returns a DArray. Is there a need for this? Can we drop it?

**Comment by malmaud** [comment]
If we're talking about enhancing `@parallel`, it might be cool to think borrowing the ['leader-follower' idea from Chapel](http://pgas11.rice.edu/papers/ChamberlainEtAl-Chapel-Iterators-PGAS11.pdf) (thanks to @jakebolewski for suggesting this). It's a good fit for Julia's philosophy of generic programming. 

In the simple case of a non-zipped range, this would mean that instead of having the scheduling logic on how work is distributed be hard-coded in the implementation of `@parallel`, it is instead up to the range. Ranges without associated "leader" logic would default to the current dynamic scheduler.

**Comment by amitmurthy** [comment]
Closing as the issue has become stale. In the event of a need expressed for `@parallel` to specify workers, we can open a new issue. `pmap` has WorkerPool now.
